hazelcast/src/main/java/com/hazelcast/internal/namespace/impl/NamespaceAwareClassLoader.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.namespace.impl;
+
+import com.hazelcast.instance.impl.Node;
+import com.hazelcast.internal.util.ExceptionUtil;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodHandles.Lookup;
+import java.lang.invoke.MethodType;
+import java.net.URL;
+import java.util.Enumeration;
+
+/**
+ * A ClassLoader that's aware of the configured namespaces and their resources.
+ * The classloading scheme does not follow the recommended {@link ClassLoader} parent delegation model: instead, this
+ * {@code ClassLoader} first looks up classes and resources on its own, then delegates if not found.
+ *
+ * @see com.hazelcast.config.NamespaceConfig
+ */
+public class NamespaceAwareClassLoader extends ClassLoader {
+    private static final MethodHandle FIND_RESOURCE_METHOD_HANDLE;
+    private static final MethodHandle FIND_RESOURCES_METHOD_HANDLE;
+
+    private final NamespaceServiceImpl namespaceService;
+    // Retain Parent for faster referencing (skips permission checks)
+    private final ClassLoader parent;
+
+    static {
+        try {
+            ClassLoader.registerAsParallelCapable();
+            Lookup lookup = MethodHandles.lookup();
+
+            FIND_RESOURCE_METHOD_HANDLE = lookup.findSpecial(ClassLoader.class, "findResource",
```

Comment:
Is the intent of these `MethodHandle`s to invoke `ClassLoader#findResource(s)` or potentially overridden versions of them in whichever classloader `pickClassLoader()` returns? I think `findSpecial` will actually find handles for the non-overridden versions, but would be good to verify intent & implementation.

Reply:
Yes you're correct - `findSpecial` will retrieve the equivalent of `super.findResource()`, rather than using overriding implementations - I don't think that's a big deal for our use-case (as demonstrated by this not failing in tests), but it's definitely an oversight. Will adjust to use `findVirtual` instead 👍

Reply:
Addressed in https://github.com/vbekiaris/hazelcast/commit/ae586bd2171025370f58cbf14cea8637633ec937